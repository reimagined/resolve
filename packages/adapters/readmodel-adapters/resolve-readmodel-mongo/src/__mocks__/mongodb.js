const documents = new Map()
let documentIndex = 1

export const result = []

export const ObjectID = id => (id != null ? String(id) : `id${documentIndex++}`)

const collection = {
  createIndex: (...args) => {
    result.push(['createIndex', ...args])
  },
  insert: (...args) => {
    result.push(['insert', ...args])
  },
  insertOne: (...args) => {
    const document = { ...args[0] }
    if (document._id == null) {
      document._id = ObjectID()
    }
    documents.set(document._id, document)
    result.push(['insertOne', document, ...args.slice(1)])
  },
  updateOne: (...args) => {
    result.push(['updateOne', ...args])
  },
  updateMany: (...args) => {
    result.push(['updateMany', ...args])
  },
  deleteMany: (...args) => {
    result.push(['deleteMany', ...args])
  },
  countDocuments: (...args) => {
    result.push(['countDocuments', ...args])
  },
  findOne: (...args) => {
    result.push(['findOne', ...args])
    if (args[0] != null && documents.has(args[0]._id)) {
      return documents.get(args[0]._id)
    }
    return {}
  },
  find: (...args) => {
    result.push(['find', ...args])
    return {
      toArray: () => []
    }
  }
}

const db = () => ({
  createCollection: (...args) => {
    result.push(['createCollection', ...args])
    return collection
  },
  collection: (...args) => {
    result.push(['collection', ...args])
    return collection
  },
  collections: (...args) => {
    result.push(['collections', ...args])
    return []
  },
  listCollections: (...args) => {
    result.push(['listCollections', ...args])
    return {
      toArray: () => []
    }
  }
})

const close = (...args) => {
  result.push(['disconnect', ...args])
  return []
}

const connect = (url, options, callback) => {
  result.push(['connect', url, options])
  const error = null
  callback(error, { db, close })
}

export const MongoClient = { connect }

ObjectID.createFromHexString = ObjectID
