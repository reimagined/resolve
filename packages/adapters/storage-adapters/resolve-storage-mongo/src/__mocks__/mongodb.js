const result = []
const database = {
  createCollection: jest.fn(),
  dropCollection: jest.fn(),
  collection: jest.fn()
}

const MongoClient = {
  connect: async (...args) => {
    result.push('connect', args)

    const client = {
      db: async (...args) => {
        result.push('db', args)

        return database
      },
      close: async (...args) => {
        result.push('close', args)
      }
    }

    return client
  }
}

export { MongoClient, database, result }
