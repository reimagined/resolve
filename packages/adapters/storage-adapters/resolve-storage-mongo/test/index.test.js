import { result as mockResult, database as mockDatabase } from 'mongodb'
import createStorageAdapter from '../src/index'

describe('resolve-storage-mongo', () => {
  let storageAdapter = null

  beforeEach(() => {
    storageAdapter = createStorageAdapter({
      collectionName: 'collectionName',
      url: 'url'
    })
  })

  afterEach(async () => {
    await storageAdapter.dispose()

    mockResult.length = 0

    jest.resetAllMocks()
  })

  test('"saveEvent" should save an event with empty payload', async () => {
    const collection = { insertOne: jest.fn() }
    mockDatabase.collection.mockImplementationOnce(async () => collection)
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))

    await storageAdapter.saveEvent({
      type: 'eventType',
      aggregateId: 'aggregateId',
      aggregateVersion: 1,
      timestamp: 1
    })

    expect(mockDatabase.collection.mock.calls).toMatchSnapshot()
    expect(collection.insertOne.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"saveEvent" should save an event', async () => {
    const collection = { insertOne: jest.fn() }
    mockDatabase.collection.mockImplementationOnce(async () => collection)
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))

    await storageAdapter.saveEvent({
      type: 'eventType',
      aggregateId: 'aggregateId',
      aggregateVersion: 1,
      timestamp: 1,
      payload: { index: 1 }
    })

    expect(mockDatabase.collection.mock.calls).toMatchSnapshot()
    expect(collection.insertOne.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"loadEvents" should load events', async () => {
    const collection = { find: jest.fn() }
    mockDatabase.collection.mockImplementationOnce(async () => collection)
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))

    const cursor = {
      sort: jest.fn().mockImplementation(() => cursor),
      project: jest.fn().mockImplementation(() => cursor),
      stream: jest.fn().mockImplementation(() => cursor),
      next: jest.fn(),
      close: jest.fn()
    }
    collection.find.mockImplementation(() => cursor)

    const loadFilters = [
      {},
      { maxEventsByTimeframe: 1 },
      { eventTypes: ['eventType'] },
      { aggregateIds: ['aggregateId'] },
      { startTime: 1, finishTime: 3 }
    ]

    for (const loadFilter of loadFilters) {
      cursor.next.mockReturnValueOnce({
        type: 'eventType',
        aggregateId: 'aggregateId',
        aggregateVersion: 1,
        timestamp: 1,
        payload: JSON.stringify(loadFilter)
      })

      cursor.next.mockReturnValueOnce(null)

      await storageAdapter.loadEvents(loadFilter, () => {})
    }

    expect(mockDatabase.collection.mock.calls).toMatchSnapshot()
    expect(cursor.sort.mock.calls).toMatchSnapshot()
    expect(cursor.project.mock.calls).toMatchSnapshot()
    expect(cursor.stream.mock.calls).toMatchSnapshot()
    expect(cursor.next.mock.calls).toMatchSnapshot()
    expect(cursor.close.mock.calls).toMatchSnapshot()
    expect(collection.find.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"getLatestEvent" should get the latest event', async () => {
    const collection = { find: jest.fn() }
    mockDatabase.collection.mockImplementationOnce(async () => collection)
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))

    const cursor = {
      sort: jest.fn().mockImplementation(() => cursor),
      project: jest.fn().mockImplementation(() => cursor),
      skip: jest.fn().mockImplementation(() => cursor),
      limit: jest.fn().mockImplementation(() => cursor),
      toArray: jest.fn()
    }
    collection.find.mockImplementation(() => cursor)

    const loadFilters = [
      {},
      { eventTypes: ['eventType'] },
      { aggregateIds: ['aggregateId'] },
      { startTime: 1, finishTime: 3 }
    ]

    for (const loadFilter of loadFilters) {
      cursor.toArray.mockReturnValueOnce([
        {
          type: 'eventType',
          aggregateId: 'aggregateId',
          aggregateVersion: 1,
          timestamp: 1,
          payload: JSON.stringify(loadFilter)
        }
      ])

      await storageAdapter.getLatestEvent(loadFilter)
    }

    expect(mockDatabase.collection.mock.calls).toMatchSnapshot()
    expect(cursor.sort.mock.calls).toMatchSnapshot()
    expect(cursor.project.mock.calls).toMatchSnapshot()
    expect(cursor.skip.mock.calls).toMatchSnapshot()
    expect(cursor.limit.mock.calls).toMatchSnapshot()
    expect(cursor.toArray.mock.calls).toMatchSnapshot()
    expect(collection.find.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"export" should return eventStream', async () => {
    const collection = { find: jest.fn() }
    mockDatabase.collection.mockImplementationOnce(async () => collection)
    mockDatabase.collection.mockImplementationOnce((name, _, cb) => cb(name))

    const cursor = {
      sort: jest.fn().mockImplementation(() => cursor),
      project: jest.fn().mockImplementation(() => cursor),
      skip: jest.fn().mockImplementation(() => cursor),
      limit: jest.fn().mockImplementation(() => cursor),
      toArray: jest.fn()
    }
    collection.find.mockImplementation(() => cursor)

    cursor.toArray.mockReturnValue([])

    const eventStream = storageAdapter.export()

    eventStream.on('data', event => {
      expect(JSON.parse(event)).toHaveProperty('payload')
      expect(JSON.parse(event)).toHaveProperty('aggregateVersion')
    })

    await new Promise(done => eventStream.on('end', done))

    expect(mockDatabase.collection.mock.calls).toMatchSnapshot()
    expect(cursor.sort.mock.calls).toMatchSnapshot()
    expect(cursor.project.mock.calls).toMatchSnapshot()
    expect(cursor.skip.mock.calls).toMatchSnapshot()
    expect(cursor.limit.mock.calls).toMatchSnapshot()
    expect(cursor.toArray.mock.calls).toMatchSnapshot()
    expect(collection.find.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })
})
