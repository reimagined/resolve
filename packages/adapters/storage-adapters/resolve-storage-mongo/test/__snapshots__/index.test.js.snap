// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve-storage-mongo "export" should return eventStream 1`] = `
Array [
  Array [
    "collectionName",
  ],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 2`] = `
Array [
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 3`] = `Array []`;

exports[`resolve-storage-mongo "export" should return eventStream 4`] = `
Array [
  Array [
    0,
  ],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 5`] = `
Array [
  Array [
    200,
  ],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 6`] = `
Array [
  Array [],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 7`] = `
Array [
  Array [],
]
`;

exports[`resolve-storage-mongo "export" should return eventStream 8`] = `
Array [
  "connect",
  Array [
    "url",
    Object {
      "useNewUrlParser": true,
      "useUnifiedTopology": true,
    },
  ],
  "db",
  Array [],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 1`] = `
Array [
  Array [
    "collectionName",
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 2`] = `
Array [
  Array [
    Object {
      "aggregateVersion": -1,
      "timestamp": -1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": -1,
      "timestamp": -1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": -1,
      "timestamp": -1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": -1,
      "timestamp": -1,
    },
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 3`] = `
Array [
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 4`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
  ],
  Array [
    0,
  ],
  Array [
    0,
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 5`] = `
Array [
  Array [
    1,
  ],
  Array [
    1,
  ],
  Array [
    1,
  ],
  Array [
    1,
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 6`] = `
Array [
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 7`] = `
Array [
  Array [
    Object {
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
    },
  ],
  Array [
    Object {
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
      "type": Object {
        "$in": Array [
          "eventType",
        ],
      },
    },
  ],
  Array [
    Object {
      "aggregateId": Object {
        "$in": Array [
          "aggregateId",
        ],
      },
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
    },
  ],
  Array [
    Object {
      "timestamp": Object {
        "$gt": 1,
        "$lt": 3,
      },
    },
  ],
]
`;

exports[`resolve-storage-mongo "getLatestEvent" should get the latest event 8`] = `
Array [
  "connect",
  Array [
    "url",
    Object {
      "useNewUrlParser": true,
      "useUnifiedTopology": true,
    },
  ],
  "db",
  Array [],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 1`] = `
Array [
  Array [
    "collectionName",
  ],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 2`] = `
Array [
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
  Array [
    Object {
      "aggregateVersion": 1,
      "timestamp": 1,
    },
  ],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 3`] = `
Array [
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
  Array [
    Object {
      "_id": 0,
    },
  ],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 4`] = `
Array [
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 5`] = `
Array [
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 6`] = `
Array [
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 7`] = `
Array [
  Array [
    Object {
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
    },
  ],
  Array [
    Object {
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
    },
  ],
  Array [
    Object {
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
      "type": Object {
        "$in": Array [
          "eventType",
        ],
      },
    },
  ],
  Array [
    Object {
      "aggregateId": Object {
        "$in": Array [
          "aggregateId",
        ],
      },
      "timestamp": Object {
        "$gt": 0,
        "$lt": Infinity,
      },
    },
  ],
  Array [
    Object {
      "timestamp": Object {
        "$gt": 1,
        "$lt": 3,
      },
    },
  ],
]
`;

exports[`resolve-storage-mongo "loadEvents" should load events 8`] = `
Array [
  "connect",
  Array [
    "url",
    Object {
      "useNewUrlParser": true,
      "useUnifiedTopology": true,
    },
  ],
  "db",
  Array [],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event 1`] = `
Array [
  Array [
    "collectionName",
  ],
  Array [
    "collectionName-freeze",
    Object {
      "strict": true,
    },
    [Function],
  ],
  Array [
    "collectionName-freeze",
    Object {
      "strict": true,
    },
    [Function],
  ],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event 2`] = `
Array [
  Array [
    Object {
      "aggregateId": "aggregateId",
      "aggregateVersion": 1,
      "payload": Object {
        "index": 1,
      },
      "timestamp": 1,
      "type": "eventType",
    },
  ],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event 3`] = `
Array [
  "connect",
  Array [
    "url",
    Object {
      "useNewUrlParser": true,
      "useUnifiedTopology": true,
    },
  ],
  "db",
  Array [],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event with empty payload 1`] = `
Array [
  Array [
    "collectionName",
  ],
  Array [
    "collectionName-freeze",
    Object {
      "strict": true,
    },
    [Function],
  ],
  Array [
    "collectionName-freeze",
    Object {
      "strict": true,
    },
    [Function],
  ],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event with empty payload 2`] = `
Array [
  Array [
    Object {
      "aggregateId": "aggregateId",
      "aggregateVersion": 1,
      "timestamp": 1,
      "type": "eventType",
    },
  ],
]
`;

exports[`resolve-storage-mongo "saveEvent" should save an event with empty payload 3`] = `
Array [
  "connect",
  Array [
    "url",
    Object {
      "useNewUrlParser": true,
      "useUnifiedTopology": true,
    },
  ],
  "db",
  Array [],
]
`;
